File: javascript/all-your-base/all-your-base.js
//
// This is only a SKELETON file for the 'All Your Base' exercise. It's been provided as a
// convenience to get you started writing code faster.
//

export const convert = (digits, in_base, out_base) => {
  if (in_base < 2) {
    throw new Error("Wrong input base");
  }

  if (out_base < 2) {
    throw new Error("Wrong output base");
  }

  if (digits.length < 1 || (digits.length > 1 && digits[0] === 0)) {
    throw new Error("Input has wrong format");
  }

  for (const digit of digits) {
    if (digit >= in_base || digit < 0) {
      throw new Error("Input has wrong format");
    }
  }

  let base = 0;
  digits.reverse().forEach(function(v, i) {
    base += v * in_base ** i;
  });

  if (base == 0) {
    return [0];
  }

  const result = [];
  let power = 0;
  while (true) {
    if (out_base ** (power + 1) > base) {
      break;
    }
    power += 1;
  }

  for (let p = power; p >= 0; p--) {
    let high = out_base ** p;
    let i = 0;

    while (true) {
      if (high * (i + 1) > base) {
        break;
      }
      i += 1;
    }

    result.push(i);
    base -= high * i;
  }

  return result;
};

File: cpp/all-your-base/all_your_base.cpp
#include "all_your_base.h"
#include <algorithm>
#include <cmath>
#include <stdexcept>

namespace all_your_base {
    std::vector<unsigned int> convert(int from, std::vector<unsigned int> digit, int to) {
        if (from < 2) {
            throw std::invalid_argument("input base must be >= 2");
        }
    
        if (to < 2) {
            throw std::invalid_argument("output base must be >= 2");
        }

        unsigned int tenth = 0;
        bool onlyZeroes = true;
        
        std::reverse(digit.begin(), digit.end());
        
        for (unsigned int pos = 0; pos < digit.size(); pos++) {
            if (digit[pos] >= (unsigned int)from) {
                throw std::invalid_argument("all digits must satisfy 0 <= d < input base");
            }

            if (digit[pos] > 0) {
                onlyZeroes = false;
            }

            tenth += digit[pos] * (unsigned int)std::pow(from, pos);
        }

        std::vector<unsigned int> result;

        if (onlyZeroes) {
            return result;
        }

        while (tenth >= (unsigned int)to) {
            unsigned int remainder = tenth % to;
            tenth = (unsigned int)floor(tenth / to);
            result.push_back(remainder);
        }

        result.push_back(tenth % to);
        std::reverse(result.begin(), result.end());
        
        return result;
    }
}  // namespace all_your_base
